homeassistant/components/control4/__init__.py

```diff
@@ -0,0 +1,230 @@
+import asyncio
+import json
+import logging
+import re
+
+from aiohttp import client_exceptions
+from pyControl4.account import C4Account
+from pyControl4.director import C4Director
+import voluptuous as vol
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import (
+    CONF_HOST,
+    CONF_PASSWORD,
+    CONF_SCAN_INTERVAL,
+    CONF_TOKEN,
+    CONF_USERNAME,
+)
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import PlatformNotReady
+from homeassistant.helpers import device_registry as dr, entity
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator
+
+from .const import (
+    CONF_ACCOUNT,
+    CONF_CONFIG_LISTENER,
+    CONF_CONTROLLER_NAME,
+    CONF_DIRECTOR,
+    CONF_DIRECTOR_ALL_ITEMS,
+    CONF_DIRECTOR_MODEL,
+    CONF_DIRECTOR_SW_VERSION,
+    CONF_DIRECTOR_TOKEN_EXPIRATION,
+    CONF_LIGHT_COLD_START_TRANSITION_TIME,
+    CONF_LIGHT_TRANSITION_TIME,
+    DEFAULT_LIGHT_COLD_START_TRANSITION_TIME,
+    DEFAULT_LIGHT_TRANSITION_TIME,
+    DEFAULT_SCAN_INTERVAL,
+    DOMAIN,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+CONFIG_SCHEMA = vol.Schema({DOMAIN: vol.Schema({})}, extra=vol.ALLOW_EXTRA)
+
+PLATFORMS = ["light"]
+
+
+async def async_setup(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Stub to allow setting up this component.
+
+    Configuration through YAML is not supported at this time.
+    """
+    hass.data.setdefault(DOMAIN, {})
+    return True
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Set up Control4 from a config entry."""
+    entry_data = hass.data[DOMAIN].setdefault(entry.entry_id, {})
+
+    config = entry.data
+    account = C4Account(config[CONF_USERNAME], config[CONF_PASSWORD])
+    try:
+        await account.getAccountBearerToken()
+    except client_exceptions.ClientError as exception:
+        _LOGGER.error("Error connecting to Control4 account API: %s", exception)
+        raise PlatformNotReady()
+    entry_data[CONF_ACCOUNT] = account
+
+    controller_name = config[CONF_CONTROLLER_NAME]
+    entry_data[CONF_CONTROLLER_NAME] = controller_name
+
+    director_token_dict = await account.getDirectorBearerToken(controller_name)
+    director = C4Director(config[CONF_HOST], director_token_dict[CONF_TOKEN])
+    entry_data[CONF_DIRECTOR] = director
+    entry_data[CONF_DIRECTOR_TOKEN_EXPIRATION] = director_token_dict["token_expiration"]
+
+    # Add Control4 controller to device registry
+    controller_href = (await account.getAccountControllers())["href"]
+    entry_data[CONF_DIRECTOR_SW_VERSION] = await account.getControllerOSVersion(
+        controller_href
+    )
+
+    result = re.search("_(.*)_", controller_name)
+    entry_data[CONF_DIRECTOR_MODEL] = result.group(1).upper()
+
+    device_registry = await dr.async_get_registry(hass)
+    device_registry.async_get_or_create(
+        config_entry_id=entry.entry_id,
+        identifiers={(DOMAIN, controller_name)},
```

Comment:
`from homeassistant.helpers import device_registry as dr`

```suggestion
        connections={(dr.CONNECTION_NETWORK_MAC, mac_address)},
```

Reply:
I'm going to retain the `identifiers` part because that identifier is used to link entities to the controller in device registry with `via_device` https://github.com/home-assistant/core/pull/37632/files#diff-d12df26ffd8ede1b79f9660721ec84faR209

Reply:
Let's change `controller_name` to `controller_unique_id` so other reviewers don't go down the path of investigating this in the future as it will save a lot of time later.

Reply:
Sure, sounds good!